File: services/modelRuntime/mossenParityBridge.ts
import { randomUUID } from 'crypto'
import type {
  MossenBetaMessage,
  MossenBetaRawMessageStreamEvent,
  MossenBetaUsage,
} from '../api/mossenSdk.js'
import type {
  CanonicalStopReason,
  CanonicalStreamEvent,
  CanonicalTurnResult,
  CanonicalUsage,
} from './canonical.js'

function toMossenUsage(usage: CanonicalUsage): MossenBetaUsage {
  return {
    cache_creation_input_tokens: 0,
    cache_read_input_tokens: 0,
    input_tokens: usage.inputTokens,
    output_tokens: usage.outputTokens,
  }
}

function toMossenStopReason(
  stopReason: CanonicalStopReason,
): MossenBetaMessage['stop_reason'] {
  if (stopReason === 'compaction') {
    return 'compaction'
  }
  if (stopReason === 'max_tokens') {
    return 'max_tokens'
  }
  if (stopReason === 'pause_turn') {
    return 'pause_turn'
  }
  if (stopReason === 'refusal') {
    return 'refusal'
  }
  if (stopReason === 'stop_sequence') {
    return 'stop_sequence'
  }
  if (stopReason === 'tool_use') {
    return 'tool_use'
  }
  return 'end_turn'
}

export function canonicalTurnToMossenMessage(
  result: CanonicalTurnResult,
  fallbackModel: string,
  messageId?: string,
): MossenBetaMessage {
  const content: MossenBetaMessage['content'] = []

  if (result.thinkingText) {
    content.push({
      signature: `synthetic-thinking:${messageId ?? randomUUID()}`,
      thinking: result.thinkingText,
      type: 'thinking',
    })
  }

  if (result.visibleText) {
    content.push({
      text: result.visibleText,
      type: 'text',
    })
  }

  for (const toolRequest of result.toolRequests) {
    content.push({
      id: toolRequest.id,
      input: toolRequest.argumentsObject,
      name: toolRequest.name,
      type: 'tool_use',
    })
  }

  if (content.length === 0) {
    content.push({
      text: '',
      type: 'text',
    })
  }

  return {
    content,
    id: messageId ?? `msg_${randomUUID()}`,
    model: fallbackModel,
    role: 'assistant',
    stop_reason: toMossenStopReason(result.stopReason),
    stop_sequence: null,
    type: 'message',
    usage: toMossenUsage(result.usage),
  }
}

export class MossenParityEventState {
  private nextIndex = 0
  private openTextIndex: number | null = null
  private openThinkingIndex: number | null = null
  private readonly toolIndices = new Map<string, number>()

  consume(
    events: Iterable<CanonicalStreamEvent>,
  ): MossenBetaRawMessageStreamEvent[] {
    const emitted: MossenBetaRawMessageStreamEvent[] = []

    for (const event of events) {
      switch (event.type) {
        case 'message_start':
          emitted.push({
            message: {
              content: [],
              id: event.messageId,
              model: event.model,
              role: 'assistant',
              stop_reason: null,
              stop_sequence: null,
              type: 'message',
              usage: toMossenUsage({ inputTokens: 0, outputTokens: 0 }),
            },
            type: 'message_start',
          })
          break
        case 'thinking_start':
          this.openThinkingIndex = this.nextIndex++
          emitted.push({
            content_block: {
              signature: '',
              thinking: '',
              type: 'thinking',
            },
            index: this.openThinkingIndex,
            type: 'content_block_start',
          })
          break
        case 'thinking_delta':
          if (this.openThinkingIndex !== null) {
            emitted.push({
              delta: {
                thinking: event.text,
                type: 'thinking_delta',
              },
              index: this.openThinkingIndex,
              type: 'content_block_delta',
            })
          }
          break
        case 'thinking_end':
          if (this.openThinkingIndex !== null) {
            emitted.push({
              index: this.openThinkingIndex,
              type: 'content_block_stop',
            })
            this.openThinkingIndex = null
          }
          break
        case 'text_start':
          this.openTextIndex = this.nextIndex++
          emitted.push({
            content_block: {
              text: '',
              type: 'text',
            },
            index: this.openTextIndex,
            type: 'content_block_start',
          })
          break
        case 'text_delta':
          if (this.openTextIndex !== null) {
            emitted.push({
              delta: {
                text: event.text,
                type: 'text_delta',
              },
              index: this.openTextIndex,
              type: 'content_block_delta',
            })
          }
          break
        case 'text_end':
          if (this.openTextIndex !== null) {
            emitted.push({
              index: this.openTextIndex,
              type: 'content_block_stop',
            })
            this.openTextIndex = null
          }
          break
        case 'tool_use_start': {
          const toolIndex = this.nextIndex++
          this.toolIndices.set(event.id, toolIndex)
          emitted.push({
            content_block: {
              id: event.id,
              input: {},
              name: event.name,
              type: 'tool_use',
            },
            index: toolIndex,
            type: 'content_block_start',
          })
          break
        }
        case 'tool_use_args_delta': {
          const toolIndex = this.toolIndices.get(event.id)
          if (toolIndex !== undefined) {
            emitted.push({
              delta: {
                partial_json: event.partialJson,
                type: 'input_json_delta',
              },
              index: toolIndex,
              type: 'content_block_delta',
            })
          }
          break
        }
        case 'tool_use_end': {
          const toolIndex = this.toolIndices.get(event.id)
          if (toolIndex !== undefined) {
            emitted.push({
              index: toolIndex,
              type: 'content_block_stop',
            })
            this.toolIndices.delete(event.id)
          }
          break
        }
        case 'message_stop':
          emitted.push({
            delta: {
              stop_reason: toMossenStopReason(event.stopReason),
              stop_sequence: null,
            },
            type: 'message_delta',
            usage: {
              input_tokens: event.usage.inputTokens,
              output_tokens: event.usage.outputTokens,
            },
          })
          emitted.push({ type: 'message_stop' })
          break
        case 'provider_error':
          break
      }
    }

    return emitted
  }
}

export function canonicalEventsToMossenEventList(
  events: Iterable<CanonicalStreamEvent>,
): MossenBetaRawMessageStreamEvent[] {
  return new MossenParityEventState().consume(events)
}

export async function* canonicalEventsToMossenEvents(
  events: AsyncIterable<CanonicalStreamEvent>,
): AsyncGenerator<MossenBetaRawMessageStreamEvent, void, void> {
  const state = new MossenParityEventState()
  for await (const event of events) {
    yield* state.consume([event])
  }
}

File: crates/mossen-agent/src/services/model_runtime/mossen_parity_bridge.rs
//! Mossen parity bridge — converts canonical types to Mossen API types and back.

use std::collections::HashMap;
use serde_json::Value;
use uuid::Uuid;

use super::canonical::{
    AssistantToolRequest, CanonicalStopReason, CanonicalStreamEvent, CanonicalTurnResult,
    CanonicalUsage,
};

/// Mossen beta message content block.
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(tag = "type")]
pub enum MossenContentBlock {
    #[serde(rename = "thinking")]
    Thinking { thinking: String, signature: String },
    #[serde(rename = "text")]
    Text { text: String },
    #[serde(rename = "tool_use")]
    ToolUse {
        id: String,
        input: HashMap<String, Value>,
        name: String,
    },
}

/// Mossen beta usage.
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct MossenBetaUsage {
    pub cache_creation_input_tokens: u64,
    pub cache_read_input_tokens: u64,
    pub input_tokens: u64,
    pub output_tokens: u64,
}

/// Mossen beta message.
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct MossenBetaMessage {
    pub content: Vec<MossenContentBlock>,
    pub id: String,
    pub model: String,
    pub role: String,
    pub stop_reason: Option<String>,
    pub stop_sequence: Option<String>,
    #[serde(rename = "type")]
    pub msg_type: String,
    pub usage: MossenBetaUsage,
}

fn to_mossen_usage(usage: &CanonicalUsage) -> MossenBetaUsage {
    MossenBetaUsage {
        cache_creation_input_tokens: 0,
        cache_read_input_tokens: 0,
        input_tokens: usage.input_tokens,
        output_tokens: usage.output_tokens,
    }
}

fn to_mossen_stop_reason(stop_reason: CanonicalStopReason) -> Option<String> {
    let s = match stop_reason {
        CanonicalStopReason::Compaction => "compaction",
        CanonicalStopReason::MaxTokens => "max_tokens",
        CanonicalStopReason::PauseTurn => "pause_turn",
        CanonicalStopReason::Refusal => "refusal",
        CanonicalStopReason::StopSequence => "stop_sequence",
        CanonicalStopReason::ToolUse => "tool_use",
        CanonicalStopReason::EndTurn => "end_turn",
    };
    Some(s.to_string())
}

/// Convert a canonical turn result into a Mossen beta message.
pub fn canonical_turn_to_mossen_message(
    result: &CanonicalTurnResult,
    fallback_model: &str,
    message_id: Option<&str>,
) -> MossenBetaMessage {
    let mut content: Vec<MossenContentBlock> = Vec::new();
    let mid = message_id
        .map(|s| s.to_string())
        .unwrap_or_else(|| format!("msg_{}", Uuid::new_v4()));

    if !result.thinking_text.is_empty() {
        content.push(MossenContentBlock::Thinking {
            signature: format!("synthetic-thinking:{}", mid),
            thinking: result.thinking_text.clone(),
        });
    }

    if !result.visible_text.is_empty() {
        content.push(MossenContentBlock::Text {
            text: result.visible_text.clone(),
        });
    }

    for tool_request in &result.tool_requests {
        content.push(MossenContentBlock::ToolUse {
            id: tool_request.id.clone(),
            input: tool_request.arguments_object.clone(),
            name: tool_request.name.clone(),
        });
    }

    if content.is_empty() {
        content.push(MossenContentBlock::Text {
            text: String::new(),
        });
    }

    MossenBetaMessage {
        content,
        id: mid,
        model: fallback_model.to_string(),
        role: "assistant".to_string(),
        stop_reason: to_mossen_stop_reason(result.stop_reason),
        stop_sequence: None,
        msg_type: "message".to_string(),
        usage: to_mossen_usage(&result.usage),
    }
}

/// State machine for converting canonical stream events to Mossen raw stream events.
pub struct MossenParityEventState {
    next_index: usize,
    open_text_index: Option<usize>,
    open_thinking_index: Option<usize>,
    tool_indices: HashMap<String, usize>,
}

impl MossenParityEventState {
    pub fn new() -> Self {
        Self {
            next_index: 0,
            open_text_index: None,
            open_thinking_index: None,
            tool_indices: HashMap::new(),
        }
    }

    /// Consume canonical stream events and produce Mossen raw stream events.
    pub fn consume(&mut self, events: &[CanonicalStreamEvent]) -> Vec<Value> {
        let mut emitted: Vec<Value> = Vec::new();

        for event in events {
            match event {
                CanonicalStreamEvent::MessageStart { message_id, model } => {
                    emitted.push(serde_json::json!({
                        "type": "message_start",
                        "message": {
                            "content": [],
                            "id": message_id,
                            "model": model,
                            "role": "assistant",
                            "stop_reason": null,
                            "stop_sequence": null,
                            "type": "message",
                            "usage": { "input_tokens": 0, "output_tokens": 0, "cache_creation_input_tokens": 0, "cache_read_input_tokens": 0 }
                        }
                    }));
                }
                CanonicalStreamEvent::ThinkingStart => {
                    let idx = self.next_index;
                    self.next_index += 1;
                    self.open_thinking_index = Some(idx);
                    emitted.push(serde_json::json!({
                        "type": "content_block_start",
                        "index": idx,
                        "content_block": { "type": "thinking", "thinking": "", "signature": "" }
                    }));
                }
                CanonicalStreamEvent::ThinkingDelta { text } => {
                    if let Some(idx) = self.open_thinking_index {
                        emitted.push(serde_json::json!({
                            "type": "content_block_delta",
                            "index": idx,
                            "delta": { "type": "thinking_delta", "thinking": text }
                        }));
                    }
                }
                CanonicalStreamEvent::ThinkingEnd => {
                    if let Some(idx) = self.open_thinking_index.take() {
                        emitted.push(serde_json::json!({
                            "type": "content_block_stop",
                            "index": idx
                        }));
                    }
                }
                CanonicalStreamEvent::TextStart => {
                    let idx = self.next_index;
                    self.next_index += 1;
                    self.open_text_index = Some(idx);
                    emitted.push(serde_json::json!({
                        "type": "content_block_start",
                        "index": idx,
                        "content_block": { "type": "text", "text": "" }
                    }));
                }
                CanonicalStreamEvent::TextDelta { text } => {
                    if let Some(idx) = self.open_text_index {
                        emitted.push(serde_json::json!({
                            "type": "content_block_delta",
                            "index": idx,
                            "delta": { "type": "text_delta", "text": text }
                        }));
                    }
                }
                CanonicalStreamEvent::TextEnd => {
                    if let Some(idx) = self.open_text_index.take() {
                        emitted.push(serde_json::json!({
                            "type": "content_block_stop",
                            "index": idx
                        }));
                    }
                }
                CanonicalStreamEvent::ToolUseStart { id, name } => {
                    let idx = self.next_index;
                    self.next_index += 1;
                    self.tool_indices.insert(id.clone(), idx);
                    emitted.push(serde_json::json!({
                        "type": "content_block_start",
                        "index": idx,
                        "content_block": { "type": "tool_use", "id": id, "name": name, "input": {} }
                    }));
                }
                CanonicalStreamEvent::ToolUseArgsDelta { id, partial_json } => {
                    if let Some(&idx) = self.tool_indices.get(id) {
                        emitted.push(serde_json::json!({
                            "type": "content_block_delta",
                            "index": idx,
                            "delta": { "type": "input_json_delta", "partial_json": partial_json }
                        }));
                    }
                }
                CanonicalStreamEvent::ToolUseEnd { id } => {
                    if let Some(idx) = self.tool_indices.remove(id) {
                        emitted.push(serde_json::json!({
                            "type": "content_block_stop",
                            "index": idx
                        }));
                    }
                }
                CanonicalStreamEvent::MessageStop { stop_reason, usage } => {
                    emitted.push(serde_json::json!({
                        "type": "message_delta",
                        "delta": {
                            "stop_reason": to_mossen_stop_reason(*stop_reason),
                            "stop_sequence": null
                        },
                        "usage": { "input_tokens": usage.input_tokens, "output_tokens": usage.output_tokens }
                    }));
                    emitted.push(serde_json::json!({ "type": "message_stop" }));
                }
                CanonicalStreamEvent::ProviderError { .. } => {}
            }
        }

        emitted
    }
}

impl Default for MossenParityEventState {
    fn default() -> Self {
        Self::new()
    }
}

/// Convert a list of canonical events into Mossen raw events (non-streaming).
pub fn canonical_events_to_mossen_event_list(events: &[CanonicalStreamEvent]) -> Vec<Value> {
    MossenParityEventState::new().consume(events)
}
